lang/resources/org/partiql/type-domains/partiql.ion

```diff
@@ -426,4 +454,279 @@
+                // part of the public API, we should consider moving this definition to the partiql_ast domain.
+                (struct parts::(* struct_part 1))
+            )
+        )
+
+        // These should be excluded as well since they were referenced only by the `select` variant of `expr`, which
+        // was excluded above.
+        (exclude
+            project_item
+            projection
+            from_source
+        )
+
+        // Change let_binding so that it has a var_decl instead of a only name to represent bindings.
+        (exclude let_binding)
+        (include (product let_binding value::expr decl::var_decl))
+
+        // Now we include new stuff, including PartiQL's relational algebra.
+        (include
+            // Every instance of `var_decl` introduces a new binding in the current scope.
+            // Every part of the AST that can introduce a variable should be represented with one of these nodes.
+            // Examples of variable declarations include:
+            // - The `AS`, `AT`, and `BY` sub-clauses in `FROM`
+            // - The `AS` sub-clauses in within a `LET` clause.
+            // - The `AS` and `AT` names specified with, `PIVOT`, i.e. `PIVOT x AS y AT z`
+            // Note that `AS` aliases specified in a select list (i.e. `SELECT x AS y` are *not* variables, they are
+            // fields.)
+            // Modeling this with a separate node (as opposed to just a symbol) is beneficial because it is easy to
+            // identify all variable declarations within a logical plan during tree traversal, and because in later
+            // permuted domains we can add information to this type such as the variable's assigned index.
+            // Elements:
+            // - `name`: the name of the variable as specified by the query author or determined statically.
+            (product var_decl name::symbol)
+
+            // The operators of PartiQL's relational algebra. See `$projectDir/docs/dev/RELATIONAL-ALGEBRA.md` for
+            // more information.  Not all operators are included here yet.
+            (sum bexpr
+                // Converts a value collection to a bindings collection.  Not used to perform physical reads.  (For
+                // that, see bexpr.project in the partiql_physical domain.)  If evaluating `expr` results in a scalar
+                // value, it is converted into a singleton bag.
+                (scan expr::expr as_decl::var_decl at_decl::(? var_decl) by_decl::(? var_decl))
+
+                // Evaluates `predicate` within the scope of every row of `bexpr`, and only returns those
+                // rows for which `predicate` returns true.
+                (filter predicate::expr source::bexpr)
+
+                // Basic join operator.  Covers cross, inner, left, right and full joins.
+                // For cross joins, set `join_type` to `(inner)` and the `predicate` to `(lit true)`.
```

Comment:
For this and the other join definition in physical_plan: curious why going w/ this approach for representing cross join? Does going w/ the following make sense?

```bash
(sum join_type (cross) (inner) (left) (right) (full))

(join
    join_type::join_type
    left::bexpr
    right::bexpr
    predicate::(? expr))
```

Reply:
I bristled when I first saw this particular modeling of cross joins as well.  This is fundamentally similar to how they are modeled with the AST today.  ...

To demonstrate why it doesn't make sense to add `(cross)` to `join_type`:  how should the evaluator behave when the `predicate` is not null and the `join_type` is `(cross)`?  

- Throw an exception?  (That would be an edge case that would have to be handled many places, which would increase the noise level in the implementation.)
- Honor the predicate anyway?  But that would be identical to an inner join, making `(cross)` redundant.

Reply:
We could ignore the predicate when the `join_type` is `cross`. Although still not ideal, having it defined explicitly is better than implicit IMO. Thinking about this further and also [looking at the grammar](https://jakewheat.github.io/sql-overview/sql-92-grammar.html#joined-table), perhaps we an also go with the following model (or something similar)?

```
(sum join
  (product left left::bexpr predicate::expr)
  (product right right::bexpr predicate::expr)
  (product inner left::bexpr right::bexpr predicate::expr)
  (product cross  left::bexpr right::bexpr))
```

Reply:
1.  `left`, `right` should have the same elements as `inner`--they each requires a left and right `bexpr` and a `predicate`.
2.  With this proposal, the code that consumes the generated classes is actually more complex.  I.e. in the `VisitorTransform`s, I now have to override 4 different methods instead of 1.  Pattern matching too has similar concerns.
3. With `inner`, I can still have still set the predicate to `(lit true)` and get something that's semantically equivalent to `cross`).  This makes `cross` kind of redundant.  Such redundancies generally increase complexity.

It's funny--I distinctly remember having this exact discussion (but related to `ExprNode`) with @almann and @therapon  that I had your position back then.  They argued strongly for modeling `cross` as an `inner` with a `(lit true)` predicate.  It's taken time but I've come around.

Reply:
For `1`, yes that was a miss from my side. For `2` we could have a wrapper type with having a member property as `kind` that references the join sum. I see the `3` orthogonal to choice of having `cross` or not as this is still the case without it. 

I hope I come around one day, but at this stage, I just can't see having an implicit logic to model rather than a precise model is a better choice. Let me know if you want to discuss offline and we can take it from there.

Reply:
From offline discussion we will make `predicate` nullable so that if the user did not specify a predicate (as with `cross join`), the AST will not lie by saying there was a `(lit true)`.

Reply:
I approved, but putting this non-blocking comment here: Although we disregard a join w/ predicate as cross join, to be more defensive, I think we could still go w/ having the `(cross_join)` as a join type to ensure the absence of predicate is not accidental to infer the cross join.